File: src/test/java/io/chesslave/eyes/SinglePieceRecognitionTest.java
package io.chesslave.eyes;

import io.chesslave.model.Color;
import io.chesslave.model.Piece;
import io.chesslave.model.Square;
import org.junit.Test;

public abstract class SinglePieceRecognitionTest extends BaseRecognitionTest {

    @Test
    public void whiteRookOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.ROOK, Color.WHITE));
    }

    @Test
    public void blackRookOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.ROOK, Color.BLACK));
    }

    @Test
    public void whiteRookOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.ROOK, Color.WHITE));
    }

    @Test
    public void blackRookOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.ROOK, Color.BLACK));
    }

    @Test
    public void whiteKnightOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.KNIGHT, Color.WHITE));
    }

    @Test
    public void blackKnightOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.KNIGHT, Color.BLACK));
    }

    @Test
    public void whiteKnightOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.KNIGHT, Color.WHITE));
    }

    @Test
    public void blackKnightOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.KNIGHT, Color.BLACK));
    }

    @Test
    public void whitePawnOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.PAWN, Color.WHITE));
    }

    @Test
    public void blackPawnOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.PAWN, Color.BLACK));
    }

    @Test
    public void whitePawnOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.PAWN, Color.WHITE));
    }

    @Test
    public void blackPawnOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.PAWN, Color.BLACK));
    }

    @Test
    public void whiteQueenOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.QUEEN, Color.WHITE));
    }

    @Test
    public void blackQueenOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.QUEEN, Color.BLACK));
    }

    @Test
    public void whiteQueenOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.QUEEN, Color.WHITE));
    }

    @Test
    public void blackQueenOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.QUEEN, Color.BLACK));
    }

    // FAIL
    @Test
    public void whiteKingOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.KING, Color.WHITE));
    }

    @Test
    public void blackKingOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.KING, Color.BLACK));
    }

    @Test
    public void whiteKingOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.KING, Color.WHITE));
    }

    @Test
    public void blackKingOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.KING, Color.BLACK));
    }

    @Test
    public void whiteBishopOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.BISHOP, Color.WHITE));
    }

    @Test
    public void blackBishopOnDarkSquare() throws Exception {
        withPieceOnSquare(Square.of("a1"), Piece.of(Piece.Type.BISHOP, Color.BLACK));
    }

    @Test
    public void whiteBishopOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.BISHOP, Color.WHITE));
    }

    @Test
    public void blackBishopOnLightSquare() throws Exception {
        withPieceOnSquare(Square.of("b1"), Piece.of(Piece.Type.BISHOP, Color.BLACK));
    }

    @Test
    public void emptyLightSquare() throws Exception {
        withEmptySquare(Square.of("b1"));
    }

    @Test
    public void emptyDarkSquare() throws Exception {
        withEmptySquare(Square.of("a1"));
    }

    public void withPieceOnSquare(Square square, Piece piece) throws Exception {
    }

    public void withEmptySquare(Square square) throws Exception {
    }
}

File: backend/src/test/java/io/chesslave/eyes/SinglePieceRecognitionTest.kt
package io.chesslave.eyes

import io.chesslave.model.Board
import io.chesslave.model.Color
import io.chesslave.model.Piece
import io.chesslave.model.Piece.Type
import io.chesslave.model.Square
import io.chesslave.visual.rendering.ChessSet
import org.junit.Test

abstract class SinglePieceRecognitionTest(chessSet: ChessSet) : BaseRecognitionTest(chessSet) {

    @Test
    fun whiteRookOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.whiteRook)
    }

    @Test
    fun blackRookOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.blackRook)
    }

    @Test
    fun whiteRookOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.whiteRook)
    }

    @Test
    fun blackRookOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.blackRook)
    }

    @Test
    fun whiteKnightOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.whiteKnight)
    }

    @Test
    fun blackKnightOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.blackKnight)
    }

    @Test
    fun whiteKnightOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.whiteKnight)
    }

    @Test
    fun blackKnightOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.blackKnight)
    }

    @Test
    fun whitePawnOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.whitePawn)
    }

    @Test
    fun blackPawnOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.blackPawn)
    }

    @Test
    fun whitePawnOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.whitePawn)
    }

    @Test
    fun blackPawnOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.blackPawn)
    }

    @Test
    fun whiteQueenOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.whiteQueen)
    }

    @Test
    fun blackQueenOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.blackQueen)
    }

    @Test
    fun whiteQueenOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.whiteQueen)
    }

    @Test
    fun blackQueenOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.blackQueen)
    }

    @Test
    fun whiteKingOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.whiteKing)
    }

    @Test
    fun blackKingOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.blackKing)
    }

    @Test
    fun whiteKingOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.whiteKing)
    }

    @Test
    fun blackKingOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.blackKing)
    }

    @Test
    fun whiteBishopOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.whiteBishop)
    }

    @Test
    fun blackBishopOnDarkSquare() {
        withPieceOnSquare(Board.a1, Piece.blackBishop)
    }

    @Test
    fun whiteBishopOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.whiteBishop)
    }

    @Test
    fun blackBishopOnLightSquare() {
        withPieceOnSquare(Board.b1, Piece.blackBishop)
    }

    @Test
    fun emptyLightSquare() {
        withEmptySquare(Board.b1)
    }

    @Test
    fun emptyDarkSquare() {
        withEmptySquare(Board.a1)
    }

    open fun withPieceOnSquare(square: Square, piece: Piece) {
    }

    open fun withEmptySquare(square: Square) {
    }
}
